flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,369 @@
+		for (int i = 0; i < sizes.length; ++i) {
+			sizes[i] = buffers[i].readableBytes();
+		}
+
+		return sizes;
+	}
+
+	private ByteBuf mergeBuffers(ByteBuf[] buffers) {
+		ByteBuf allData = ALLOCATOR.buffer();
+		for (ByteBuf buffer : buffers) {
+			allData.writeBytes(buffer);
+		}
+
+		return allData;
+	}
+
+	private List<ByteBuf> partitionBuffer(ByteBuf buffer, int[] sizes) {
+		// For each group of three messages [A, B, C], we repartition them as three buffers:
+		//  1. First half of A.
+		//  2. Second half of A + B + first half of C.
+		//  3. Second half of C.
+		List<Integer> partitionPositions = new ArrayList<>();
+
+		int position = 0;
+		partitionPositions.add(position);
+
+		for (int i = 0; i < sizes.length / 3 * 3; i += 3) {
+			position += sizes[i] / 2;
+			partitionPositions.add(position);
+
+			position += (sizes[i] / 2 + sizes[i + 1] + sizes[i + 2] / 2);
+			partitionPositions.add(position);
+
+			position += sizes[i + 2] / 2;
+			partitionPositions.add(position);
+		}
+
+		// Partitions the remaining data into three parts averagely.
+		int remainingBytes = buffer.readableBytes() - position;
+
+		position += (remainingBytes / 3);
+		partitionPositions.add(position);
+
+		position += (remainingBytes / 3 * 2);
+		partitionPositions.add(position);
+
+		partitionPositions.add(buffer.readableBytes());
+
+		// Partitions the buffers according to the selected partition positions.
+		List<ByteBuf> result = new ArrayList<>();
+		for (int i = 1; i < partitionPositions.size(); ++i) {
+			ByteBuf partitionedBuffer = ALLOCATOR.buffer();
+			partitionedBuffer.writeBytes(
+				buffer,
+				partitionPositions.get(i - 1),
+				partitionPositions.get(i) - partitionPositions.get(i - 1));
+			result.add(partitionedBuffer);
+		}
+
+		return result;
+	}
+
+	private List<NettyMessage> decodedMessages(EmbeddedChannel channel, List<ByteBuf> inputBuffers) {
+		for (ByteBuf buffer : inputBuffers) {
+			channel.writeInbound(buffer);
+		}
+
+		channel.runPendingTasks();
+
+		List<NettyMessage> decodedMessages = new ArrayList<>();
+		Object input;
+		while ((input = channel.readInbound()) != null) {
+			assertTrue(input instanceof NettyMessage);
+			decodedMessages.add((NettyMessage) input);
+		}
+
+		return decodedMessages;
+	}
+
+	private void verifyDecodedMessages(
+		List<NettyMessage> expectedMessages,
+		List<NettyMessage> decodedMessages,
+		InputChannelID normalChannelId) {
+
+		assertEquals(expectedMessages.size(), decodedMessages.size());
+		for (int i = 0; i < expectedMessages.size(); ++i) {
+			assertEquals(expectedMessages.get(i).getClass(), decodedMessages.get(i).getClass());
+
+			if (expectedMessages.get(i) instanceof NettyMessage.BufferResponse) {
+				BufferResponse expected = (BufferResponse) expectedMessages.get(i);
+				BufferResponse actual = (BufferResponse) decodedMessages.get(i);
+
+				verifyBufferResponseHeader(expected, actual);
+				if (expected.bufferSize == 0 || !expected.receiverId.equals(normalChannelId)) {
```

Comment:
better to pass `releasedChannelId` in the argument instead.

Reply:
I think `releasedChannelId` is not required since we already compared expected receiver id with the actual receiver id, thus the actual receiver id should always be right. Then we could distinguish normal ones with other two ids.

If we pass `releasedChannelId`, we would also have to pass `removedChannelId`.